htlcswitch/link.go

```diff
@@ -2369,10 +2369,24 @@ func (l *channelLink) processRemoteSettleFails(fwdPkg *channeldb.FwdPkg,
 				outgoingHTLCID: pd.ParentIndex,
 				destRef:        pd.DestRef,
 				htlc: &lnwire.UpdateFailHTLC{
-					Reason: lnwire.OpaqueReason(pd.FailReason),
+					Reason: lnwire.OpaqueReason(
+						pd.FailReason,
+					),
 				},
 			}
 
+			// If the failure message lacks an HMAC (but includes
+			// the 4 bytes for encoding the message and padding
+			// lengths, then this means that we received it as an
+			// UpdateFailMalformedHTLC. As a result, we'll signal
+			// that we need to convert this error within the switch
+			// to an actual error, by encrypting it as if we were
+			// the originating hop.
+			convertedErrorSize := lnwire.FailureMessageLength + 4
+			if len(pd.FailReason) == convertedErrorSize {
```

Comment:
This looks like a short cut, checking the length to find out if it is encrypted or not. Isn't there a cleaner way?

If was thinking of encrypting the error already in `handleDownStreamPkt` when the `UpdateFailMalformedHTLC` comes in. Just calling `EncryptFirstHop` there.

Reply:
Maybe that doesn't work if we are the sender.

Reply:
Currently we always let the switch handle the onion encryption for errors as it holds the circuit which holds the shared secret. 

One alternative path is to modify the channel state machine itself to differentiate between a regular fail, and a malformed fail. We'd then need to also store the malformed fail message on disk within commitment diffs as well as forwarding packages in the database. Additionally, we'd need a new message within the channel state machine as well to maintain the 1:1 mapping between what's stored and the contents of the update log. Overall, it's a much bigger change than what's being proposed in this PR as it'll eliminate this new if statement from this diff, implement the above, then also modify the link to understand this message. 

I'm fond of the fix described above as IMO it implements a more robust solution (recognizes the scenario rather than relying on a length check), however it would increase the scope of this PR and surface area of modified code. If we like the alternative fixed described above, I propose we move forward with it, but also merge this as is since it fixes an existing issue and is small enough to be included in the next release cycle.

Reply:
And how about fixing it like this? (last commit)
https://github.com/lightningnetwork/lnd/compare/master...joostjager:new-onion-structs-alt

To me, it seems that fixing the malformed htlc case is not necessarily needed in 0.6.1. Unless I miss an urgent reason to do it now, I think it should be fixed properly without the len check.

Reply:
That diff is incomplete, it only implements part of the change I mentioned above. We must fix it now as otherwise the integration tests fail.

Reply:
From the change you described:

1. Modify the channel state machine itself to differentiate between a regular fail, and a malformed fail. 
2. We'd then need to also store the malformed fail message on disk within commitment diffs as well as forwarding packages in the database.
3. Additionally, we'd need a new message within the channel state machine as well to maintain the 1:1 mapping between what's stored and the contents of the update log. 

Looking through `channel.go`, it does seem that most of this is already in. Which bigger change is still remaining after this diff?

Reply:
With regards to the failing itest: we could postpone fixing the expected message in the itest.

Reply:
Fixing the expected message without fixing handling of malformed errors won't resolve anything.